packages/core/src/rules/common/no-required-schema-properties-undefined.ts

```diff
@@ -0,0 +1,82 @@
+import { Oas2Rule, Oas3Rule } from '../../visitors';
+import { Oas3Schema, Oas3_1Schema } from '../../typings/openapi';
+import { Oas2Schema } from 'core/src/typings/swagger';
+import { UserContext } from 'core/src/walk';
+import { isRef } from '../../ref-utils';
+
+export const NoRequiredSchemaPropertiesUndefined: Oas3Rule | Oas2Rule = () => {
+  const processNestedSchemasRecursively = (
+    schema: Oas3Schema | Oas3_1Schema | Oas2Schema,
+    { resolve }: UserContext,
+    allProperties: Record<string, Oas3Schema | Oas3_1Schema | Oas2Schema>,
+    visitedSchemas: Set<Oas3Schema | Oas3_1Schema | Oas2Schema>
+  ) => {
+    // Skip oneOf/anyOf as it's complicated to validate it right now.
+    // We need core support for checking constraints through those keywords.
+    // Right now we only resolve allOf keyword nesting.
+    if (schema.allOf) {
+      // Check if the schema has been visited before processing it
+      if (visitedSchemas.has(schema)) {
+        return;
+      }
+
+      // Add the schema to the visited set to avoid circular references
+      visitedSchemas.add(schema);
+
+      for (const nestedSchema of schema.allOf) {
+        if (isRef(nestedSchema)) {
+          const resolved = resolve(nestedSchema).node as Oas3Schema | Oas3_1Schema | Oas2Schema;
+          Object.assign(allProperties, resolved.properties);
```

Comment:
Could you try rewriting the function to avoid the mutation on different levels? What we really want, is to return all the properties from the nested level to the parent and merge with it's properties.

Could you briefly check if something like the following would also work? 

```ts
const elevateProperties = schema => { 
  if (isRef(schema)) { 
    return elevateProperties(resolve(schema).node)
  }
  return Object.assign({}, schema.properties, ...schema.allOf?.map(elevateProperties))
}
```

Reply:
Refactored and fixed, thanks!